### PR TITLE
Add change_detection to only detect changes when True

### PR DESCRIPTION
**Pull request by dylanShark**
Add an optional flag change_detection. If it is not set, pomgen detects changes and will only generate pom if changes detected. If change_detection is explicitly set to False, pomgen would always generate pom no matter if there is change to release.
pomgen query result would be marked ` ! always released (change detection is disabled)`.

### CODE REVIEW

**Comment by simontoens** [comment]
Hi @dylanShark 

I think these 2 methods may need to get updated:
https://github.com/salesforce/pomgen/blob/master/crawl/crawler.py#L462
https://github.com/salesforce/pomgen/blob/master/crawl/libaggregator.py#L126

Also, since we already have the FORCE release reason, maybe we can re-use that one for this?
https://github.com/salesforce/pomgen/blob/master/crawl/releasereason.py#L13

**Comment by dylanShark** [comment]
> Hi @dylanShark
> 
> I think these 2 methods may need to get updated:
> https://github.com/salesforce/pomgen/blob/master/crawl/crawler.py#L462
> https://github.com/salesforce/pomgen/blob/master/crawl/libaggregator.py#L126
> 
> Also, since we already have the FORCE release reason, maybe we can re-use that one for this (then probably the 2 methods above don't need to be updated)?
> https://github.com/salesforce/pomgen/blob/master/crawl/releasereason.py#L13

Thanks for the review! I was looking into the 2 methods you pointed and yes they need to be updated. Sorry I did not catch this in testing because I tested my changes using a library node with single artifact, so [test case 1](https://github.com/salesforce/pomgen/blob/master/crawl/crawler.py#L381) and [test case 2](https://github.com/salesforce/pomgen/blob/master/crawl/libaggregator.py#L128) were not tested. I will be sure to add test for these cases as well. 
I think because FORCE is set via command line option `--force` instead of in the BUILD.pom, reusing that may send an unclear message to the user? Also pomgen query gets the release reasons from [here](https://github.com/salesforce/pomgen/blob/master/crawl/libaggregator.py#L82), FORCE is not a valid case because `--force` is not a valid commandline argument for `@pomgen//:query`. So I think we should keep the ALWAYS release reason as a separate case.

**Comment by simontoens** [comment]
Hi @dylanShark, I think it doesn't really make sense to have both release reasons "ALWAYS" and "FORCE".  Lets just keep "ALWAYS", that sounds better anyway.  I think more or less you only need to change this line to use ALWAYS instead of FORCE: https://github.com/salesforce/pomgen/blob/master/crawl/crawler.py#L333.  So if the user runs pomgen with --force, then the release reason for all artifacts becomes "ALWAYS".  

You are right, --force is only supported by pomgen, not query.  We could easily add it to query too - it wouldn't really provide much value, but it would be useful for testing (when running query with --force, all artifacts should be marked as needing to be released).  To add it to query, just pass it through here: https://github.com/salesforce/pomgen/blob/master/query_maven_metadata.py#L158.

File: tests/buildpomtest.py [hunk 1]
```diff
@@ -32,11 +32,16 @@ def test_parse_BUILD_pom(self):
         self.assertIs(pomgenmode.DYNAMIC, art_def.pom_generation_mode)
         self.assertEqual(None, art_def.pom_template_file)
         self.assertTrue(art_def.include_deps)
+        self.assertTrue(art_def.change_detection)
         self.assertEqual(package_rel_path, art_def.bazel_package)
         self.assertEqual(None, art_def.released_version)
         self.assertEqual(None, art_def.released_artifact_hash)
         self.assertIsNotNone(art_def.version_increment_strategy)
 
+        self._write_build_pom_set_change_detection(repo_package, artifact_id, group_id, version, "dynamic", False)
```
Comment (simontoens) [inline tests/buildpomtest.py:41]: @dylanShark just make this a different test please (new test method).  Since we can, lets add 2 new test methods.  One that sets change_detection explicitly to True and one that sets change_detection explicitly to False.

Comment (dylanShark) [reply tests/buildpomtest.py:41]: Will do!

File: query_maven_metadata.py [hunk 2]
```diff
@@ -61,6 +61,8 @@ def _parse_arguments(args):
     # this is experimental and not generalized yet
     parser.add_argument("--filter", type=str, required=False,
         help="Generic query filter, currently only supported for artifact queries")
+    parser.add_argument("--force", required=False, action='store_true',
+        help="If set, always generated poms, regardless of whether an artifact has changed since it was last released")
```
Comment (simontoens) [inline query_maven_metadata.py:65]: @dylanShark this help msg doesn't make sense, since query doesn't generate poms.  Could you rephrase please?

Comment (dylanShark) [reply query_maven_metadata.py:65]: @simontoens I will update it, sorry I missed the help message.

File: crawl/releasereason.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 class ReleaseReason(object):
     FIRST = "artifact has never been released"
     ARTIFACT = "binary artifact changed"
+    ALWAYS = "always released (--force option used or change detection disabled)"
```
Comment (simontoens) [inline crawl/releasereason.py:11]: @dylanShark lets remove this bit in parens.  "always released" is good.

**Review by simontoens — APPROVED** [review]
